### ISSUE TITLE
1.90.0 regression: type inference fails when comparing `CString` array with empty array literal

### ISSUE THREAD

**Issue by magicant** [rust-lang/rust#146782]
### Code

I tried this code:

```rust
fn main() {
    let v: [std::ffi::CString; 0] = [];
    assert_eq!(v, []);
}
```

I expected to see this happen: successful build.

Instead, this happened: 

```
$ cargo +stable r
   Compiling e0282 v0.1.0 (/home/magicant/Projects/e0282)
error[E0282]: type annotations needed
 --> src/main.rs:3:5
  |
3 |     assert_eq!(v, []);
  |     ^^^^^^^^^^^^^^^^^ cannot infer type
  |
  = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0282`.
error: could not compile `e0282` (bin "e0282") due to 1 previous error
```

### Version it worked on

It most recently worked on: 1.89.0

### Version with regression

`rustc --version --verbose`:
```
rustc 1.90.0 (1159e78c4 2025-09-14)
binary: rustc
commit-hash: 1159e78c4747b02ef996e55082b704c09b970588
commit-date: 2025-09-14
host: x86_64-unknown-linux-gnu
release: 1.90.0
LLVM version: 20.1.8
```

### Backtrace

None. (No crash)

### Additional notes

This might be related to the new `PartialEq` impls for `CStr(ing)` that were stabilized in 1.90.0.

**Comment by theemathas** [comment]
The [crater run](https://github.com/rust-lang/rust/pull/137268#issuecomment-2846775335) did not detect yash's regression, as both `yash-builtin-0.7.0` and `yash-env-0.7.0` produced the following error in crater:

> this task or one of its parent failed: No space left on device (os error 28)

The crater run had an unusually high number of errors (152539 of them). Maybe this needs a crater rerun?

**Comment by bjoernager** [comment]
Behaviour using `CString` is now identical to how using `String` and/or `OsString` has been since rustc 1.0.0. What exactly is the problem?

**Comment by magicant** [comment]
I'm personally fine with adding type annotations on my side.  
I created this issue because I was not sure whether this incompatibility was intentional or an oversight.

**Comment by workingjubilee** [comment]
Yes, type inference is one of those things we don't promise works exactly the same, even on the algorithm level, so this is always a permitted breakage, in a sense. We usually try to avoid having it impact common cases, but our tooling for doing so is imperfect.
